dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt

```diff
@@ -78,13 +79,50 @@ private class SymbolSampleAnalysisEnvironment(
         return SampleSnippet(imports, body)
     }
 
+    // copy-pasted from StdLib 1.5
+    // TODO use it from Stdlib
+     private inline fun <T, R : Any> Iterable<T>.firstNotNullOfOrNull(transform: (T) -> R?): R? {
+        for (element in this) {
+            val result = transform(element)
+            if (result != null) {
+                return result
+            }
+        }
+        return null
+    }
+
+    // TODO: remove after KT-53669 and use [org.jetbrains.kotlin.analysis.api.symbols.sourcePsiSafe] from Analysis API
+    private inline fun <reified PSI : PsiElement> KtSymbol.kotlinAndJavaSourcePsiSafe(): PSI? {
+        // TODO: support Java sources after KT-53669
+        val sourcePsi = when (origin) {
+            KtSymbolOrigin.SOURCE -> this.psi
+            KtSymbolOrigin.JAVA -> this.psi
+
+            KtSymbolOrigin.SOURCE_MEMBER_GENERATED -> null
+            KtSymbolOrigin.LIBRARY -> null
+            KtSymbolOrigin.SAM_CONSTRUCTOR -> null
+            KtSymbolOrigin.INTERSECTION_OVERRIDE -> null
+            KtSymbolOrigin.SUBSTITUTION_OVERRIDE -> null
+            KtSymbolOrigin.DELEGATED -> null
+            KtSymbolOrigin.JAVA_SYNTHETIC_PROPERTY -> null
+            KtSymbolOrigin.PROPERTY_BACKING_FIELD -> null
+            KtSymbolOrigin.PLUGIN -> null
+            KtSymbolOrigin.JS_DYNAMIC -> null
+        }
+
+        return sourcePsi as? PSI
+    }
+
     private fun findPsiElement(sourceSet: DokkaSourceSet, fqLink: String): PsiElement? {
-        val ktSourceModule = samplesKotlinAnalysis.getModuleOrNull(sourceSet)
-            ?: projectKotlinAnalysis.getModule(sourceSet)
+        // fallback to default roots of the source set even if sample roots are assigned
```

Comment:
good idea to add a comment! it explains the code below, but it doesn't explain the intention, which is the less obvious part :)

```suggestion
        // fallback to default roots of the source set even if sample roots are assigned,
        // because `@sample` tag can contain links to local project functions
```

Reply:
Can you elaborate on what `local project functions` means? Why is it "local"?  I did not understand.

Reply:
I understood the problem and the fix the following way:

If you have the code like

```kotlin
/src/kotlin/com/example/MyFile.kt

/**
 * @sample bar
 */
fun foo() {
    
}

fun bar() {}
```

where `@sample` references a function declared in the project sources itself, like literally the function below it, we need to use `projectKotlinAnalysis`. This is what I meant by a "local" function.

And if you store sample functions outside of the project sources, like in a separate directory somewhere, and you configure it using `DokkaSourceSet#samples`, then it needs to use `samplesKotlinAnalysis`. I'd consider this "external"

The problem was that it was always using `samplesKotlinAnalysis`, which pointed to external sample functions, but it needs to resolve "local" project functions as well

If I got the idea and the fix correctly, you can use whatever words you find more suitable, that was just a suggestion :) But I might've misunderstood it, in which case it definitely needs some sort of a comment :)

Reply:
I have changed a little to avoid a "local" function.
```kt
// because `@sample` tag can contain links to functions from project sources
```